### ISSUE TITLE
Inception v3 implementation, High start val loss

### ISSUE THREAD

**Issue by WilliamJuel** [NVIDIA/caffe#298]
I have attached my prototxt for the implementation of inception v3. 
[inceptionv3_prototxt.txt](https://github.com/NVIDIA/caffe/files/844500/inceptionv3_prototxt.txt)

I am on the caffe-0.15 branch

Training goes fine and i can see the accuracy on the validation set aswell. My problem occurs when i try test a single image in DIGITS and also doing a regular forward pass with some python code, to see the weights etc. 

When i do this i get the following error: 

ERROR: error code -6 
message WARNING: Logging before InitGoogleLogging() is written to STDERR
F0315 13:38:00.334692  3672 insert_splits.cpp:29] Unknown bottom blob 'bn1' (layer 'conv1/3x3_s1_scale', bottom index 0)

Any ideas how to fix this issue, so i can use the model to classify new images.

Thanks alot.
Best regards,
William

**Comment by WilliamJuel** [comment]
I fixed the issue by replacing all the Batchnorm layers with (names changed accordingly):

`layer {
  name: "bn1"
  type: "BatchNorm"
  bottom: "conv1/3x3_s2"
  top: "bn1"
  param {
    lr_mult: 1
  }
  batch_norm_param {
    scale_filler {
      type: "constant"
      value: 1
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}`

So that i can now visualize the weights in DIGITS on a single image. 
But a new problem occurred.

The problem i am experiencing now is that the validation loss explodes in the beginning while the training loss is normal, for a 5 class problem (around 1.8). Before i changed the Batchnorm layers to the above the validation loss were the "same" as the training loss. So i speculate this has something to do with the way i have written the batchnorm layer in my prototxt? Can anyone give me an example of how to write it for the caffe-0.15 branch or give other suggest to what could be going wrong.
I will put my [solver_prototxt.txt](https://github.com/NVIDIA/caffe/files/845233/solver_prototxt.txt) here aswell

Thanks alot, 
Best regards,
William

**Comment by borisgin** [comment]
This is example how you should define BN layer for nvidia/caffe-0.15
```
layer {
  name: "bn"
  type: "BatchNorm"
  bottom: "pool"
  top: "bn"
  batch_norm_param {
    moving_average_fraction: 0.98
    use_global_stats: false
    eps: 0.0001
    scale_filler {
      type: "constant"
      value: 1 
    }
    bias_filler {
      type: "constant"
      value: 0 
    }
  }
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 1
    decay_mult: 1
  }  
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }  
}
```

**Comment by WilliamJuel** [comment]
Thank you for the reply @borisgin , i will change all my BN layers to this. This might be a stupid question, and i apologies for that, but does this mean i have to remove the adjacent scale layers like:

layer {
  name: "conv1/3x3_s1_scale"
  type: "Scale"
  bottom: "bn1"
  top: "bn1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 1
    decay_mult: 0
  }
  scale_param {
    axis: 1
    num_axes: 1
    filler {
      type: "constant"
      value: 1.0000001
    }
    bias_filler {
      type: "constant"
      value: 0.0
    }
    bias_term: true
  }

Since scaling happens in the BN layer?

**Comment by borisgin** [comment]
BatchNorm in nvcaffe already has scale and bias inside so you don't need extra scaling layer. 
Notice that BN in nvcaffe-0.15 has a constraint: top should be different than bottom ( it is removed in the next nvcaffe)

**Comment by WilliamJuel** [comment]
Thank you for the reply again @borisgin , I removed all the adjacent scale layers and wrote BN as u said. My top and bottom in all my BN layers are different already, so thats good. 

I however still have an issue with the validation loss starting really high as seen in the screenshot, is this just a visual bug or might there be some error in my net that causes this ? 

![screenshot from 2017-03-16 16-36-47](https://cloud.githubusercontent.com/assets/26434742/24004646/236546a4-0a67-11e7-82a9-31828161dace.png)

here is the prototxt i use now:
[inception_v3.prototxt.txt](https://github.com/NVIDIA/caffe/files/848236/inception_v3.prototxt.txt)

**Comment by WilliamJuel** [comment]
I changed the batchnorm as you said @borisgin, i still have the above issue with the validation loss, sadly. When i look into the visualizations from digits, specifically the BN layer it looks like this:

![bn](https://cloud.githubusercontent.com/assets/26434742/24100822/1afc2ff2-0d76-11e7-9a02-cd1cc7b583c6.png)

I do not get the desired 0 mean and 1 std. What could the reason for this be?

**Comment by borisgin** [comment]
Cudnn BN layer has scale and bias layer integrated inside.

**Comment by WilliamJuel** [comment]
Yes, i have already changed all my BN layers to what you told me @borisgin  and i removed the scale layers i had in the first place aswell. So these two problems: The high val loss and that i dont get 0 mean 1 std in BN layer (visualization in my last post) is after i changed to what you told me too :)

**Comment by WilliamJuel** [comment]
Just to clarify the confusion, this is how my BM layer looks now which i follow by a PReLU.

layer {
   name: "bn1"
   type: "BatchNorm"
   bottom: "conv1/3x3_s2"
   top: "bn1"
  batch_norm_param {
    moving_average_fraction: 0.98
    use_global_stats: false
    eps: 0.0001
    scale_filler {
      type: "constant"
      value: 1 
    }
    bias_filler {
      type: "constant"
      value: 0 
    }
  }
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 1
    decay_mult: 1
  }  
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }  
}

**Comment by borisgin** [comment]
This is how BN layer works
1. compute m and v ober batch
2.  compute nomralized  input 
     Y= (X - m) / sqrt(e + v)    
// this is intrenal blob . at that moment Y  has mean ~ 0 and var ~ 1
3. apply scale and bias to each channel:
    Y [c] = Y[c] * s [c] + b[c]
So output is shifted from 0 and its variance != 1

**Comment by WilliamJuel** [comment]
thank you for clarifying that,  then what i see makes sense; the mean is close to 0 and the var != 1 throughout the network in the Batch norm layers.

**Comment by WilliamJuel** [comment]
Hey, @borisgin i successfully trained my net with using the batchnorm you told me was the correct for nvidia/caffe-0.15. But when i try to use the model outside of digits, as an example using the classification example in the digits classification folder example.py i get the following:

python example.py snapshot_iter_47160.caffemodel deploy.prototxt 217_left.jpeg --mean mean.binaryproto

"F0419 13:36:36.483314 10905 net.cpp:804] Cannot copy param 0 weights from layer 'xbn_conv1'; shape mismatch. Source param shape is 1 64 1 1 (64); target param shape is 64 (64). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer."

Do you have a suggestion to how i can use my caffemodel with the BN layers to classify images outside of digits ? 

Thanks in advance.

**Comment by WilliamJuel** [comment]
I did not find any solution or answer to why my val loss starts insanely high. But im closing this since im not using the same specs anymore..
